### PR TITLE
Fix enabling of fanotify/fsnotify.

### PR DESCRIPTION
**Pull request by jankratochvil**
It just never could get enabled.

### CODE REVIEW

**Comment by ericcurtin** [comment]
Thanks very much for the contribution @jankratochvil !

This fails CI, it's not immediately obvious to be why, would need to be fixed before merged, this script:

https://github.com/inotify-tools/inotify-tools/blob/master/build_and_test_multi.sh

re-creates what the CI does.

**Comment by amir73il** [comment]
I am missing context.
Are you saying that fanotify was never enabled and tests never ran?
This sounds strange to me.
I am pretty sure that the fanotify tests were running and pass when we merged fanotity support.
@ericcurtin is that not the case?

**Comment by ericcurtin** [comment]
I thought so, I haven't had a moment to look at this in detail yet.

**Comment by jankratochvil** [comment]
Since the initial commit 0971a971bb05287efae1aea9a2a28bdbc3351f78 without an explicit `configure --enable-fanotify` the fanotify was disabled. So the testing scripts did never test it.
And in this commit there was `build_and_test.sh` but not `build_and_test_multi.sh`.
When `build_and_test_multi.sh` got introduced fanotify was not being tested as being disabled by default.
And fanotify for some reason (kernel bug? I did not investigate) does not work in containers.

**Comment by jankratochvil** [comment]
rerunnning the checks

**Comment by jankratochvil** [comment]
I think the existing errors are false positives only.

**Comment by amir73il** [comment]
@jankratochvil what do you mean by "fanotify does not work in containers"?
it should definitely work in privileged containers, which is what build_and_test_multi.sh uses.
if you see otherwise please report a bug with more details.
It looks to me like the tests are failing for one of the reasons below:

1. If the test is run as root, mount_filesystem is missing mkfs.$fstype -F img (to force mkfs on image file)
2. If the test is not run as root, the check if fanotify_supported --filesystem never gets to check privileges as the helper claims to do. the --filesystem needs to check if the user running the test is root

I have a vague memory about @ericcurtin running the tests manually with sudo before merging, because the CI does not run them as root, but I may be wrong.

In any case, the tests fsnotifywait --fanotify (without --filesystem) should be able to run and pass inside or outside of containers.

I have a pretty simple patch to allow fanotify --filesystem watch inside rootless containers (https://github.com/amir73il/linux/commit/2e6f12e327f88a523df87ea128c8c98cbba9fd02) but that support is not upstream and nobody asked me to expedite it.

If rootless fanotify --filesystem support would be merged, the test would have to change to mount a tmpfs
Not sure why I chose ext2 for the test - possibly so the test could run on kernels before support for fanotify+tmpfs (https://github.com/amir73il/linux/commit/59cda49ecf6c9a32fae4942420701b6e087204f6)

File: build_and_test.sh [hunk 1]
```diff
@@ -83,6 +83,8 @@ if command -v apt; then
 elif command -v apk; then
   apk add build-base alpine-sdk autoconf automake libtool bash coreutils clang \
     clang-extra-tools lld linux-headers curl git zip gcompat
+elif command -v dnf; then
+  $pre dnf install -y gcc-c++ autoconf automake doxygen make libtool clang curl git unzip
```
Comment (ericcurtin) [inline build_and_test.sh:87]: Thanks for enabling a Fedora build, meant to get around to this.

**Comment by ericcurtin** [comment]
In case you guys haven't noticed, I ported this codebase to C++ also, but I plan on using a restricted set of C++ the "-nodefaultlibs -lc -fno-exceptions" kind so we don't gain any extra runtime dependencies.

**Comment by jankratochvil** [comment]
I am running everything as root (in KVM) and I get:
```
Couldn't watch test-symlink: Operation not supported
fanotify_mark(3, FAN_MARK_ADD|FAN_MARK_DONT_FOLLOW, FAN_ACCESS|FAN_MODIFY|FAN_ATTRIB|FAN_CLOSE_WRITE|FAN_CLOSE_NOWRITE|FAN_OPEN|FAN_MOVED_FROM|FAN_MOVED_TO|FAN_CREATE|FAN_DELETE|FAN_DELETE_SELF|FAN_MOVE_SELF|FAN_ONDIR|FAN_EVENT_ON_CHILD, AT_FDCWD, "test-symlink") = -1 EOPNOTSUPP (Operation not supported)
```
It has been tested on:
- FAIL kernel-6.3.12-200.fc38.x86_64
- FAIL kernel-6.5.0-0.rc1.20230711git3f01e9fed845.12.fc39.x86_64
The second kernel should already include: [fanotify: support reporting non-decodeable file handles](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a95aef69a740f5a1c7d70f2b58552207edaef99a)

So I do not know what else to blame than kernel.

**Comment by amir73il** [comment]
Ah it's overlayfs ? This needs more code that is not upstream https://github.com/amir73il/linux/commits/ovl_encode_fid

Can you run the test on a directory that is shared from the host and not on overlayfs? Or mount a filesystem for running the test.
You can use the same helper that is used to mount ext2 fs for --filesystem test.

**Comment by jankratochvil** [comment]
> Ah it's overlayfs ?

Yes.

> Can you run the test on a directory that is shared from the host and not on overlayfs? Or mount a filesystem for running the test. You can use the same helper that is used to mount ext2 fs for --filesystem test.

That is what happens by default when using containers. Maybe it could also run a test in non-containerized mode to have better testing coverage? Personally I just wanted to enable the fanotify functionality when you did already implement it. I have tried to modify `.github/workflows/build.yml` but it did disable all the tests.

**Comment by ericcurtin** [comment]
Are there container filesystems we can use that supports this other than overlayfs? If no, we can run outside containers, just like to do everything in containers to be able to easily migrate CI or run locally etc.

**Comment by amir73il** [comment]
> Are there container filesystems we can use that supports this other than overlayfs? If no, we can run outside containers, just like to do everything in containers to be able to easily migrate CI or run locally etc.

There are several way to fix this. I will see if we can  generalize the test later

File: t/fanotify-common.sh [hunk 2]
```diff
@@ -2,7 +2,14 @@
 
 # Check for kernel support and privileges
 fanotify_supported() {
-    ../../src/fsnotifywait --fanotify $* 2>&1 | grep -q 'No files specified'
+    if [ -z "$(grep 'kthreadd' /proc/2/status 2>/dev/null)" ]; then
+        # FIXME: fanotify is broken in containers
+        # https://stackoverflow.com/a/72136877/2995591
```
Comment (amir73il) [inline t/fanotify-common.sh:7]: Please be accurate. It is not working on overlayfs.
It would be pretty easy to fix by using tmpfs but I can do that later.

**Review by amir73il — CHANGES_REQUESTED** [review]
I think the minimum fix needed for enable fanotify is to add mkfs -F flag and to test root user as requirement for --filesystem watch test

**Comment by amir73il** [comment]
> > Are there container filesystems we can use that supports this other than overlayfs? If no, we can run outside containers, just like to do everything in containers to be able to easily migrate CI or run locally etc.
> 
> There are several way to fix this. I will see if we can generalize the test later

@ericcurtin I created an improved version of fanotify_supported(). please see https://github.com/amir73il/inotify-tools/commits/fsnotify. This branch includes the fanotify build fixes by @jankratochvil, but not his Fedora fixes which are unrelated to fanotify

**Comment by ericcurtin** [comment]
@amir73il cool, feel free to open a PR for that, that one looks looks good to me. This one does too! Merging...

**Review by ericcurtin — APPROVED** [review]

**Pull request merged**